### PR TITLE
Fix online softmax epilogue fusion for T5

### LINKED ISSUE #177118: [Inductor]T5 models performance dropped 10% on inductor mode after pattern match changes
### 🐛 Describe the bug

T5 models' performance dropped 10% in inductor mode after PR https://github.com/pytorch/pytorch/commit/d50721ab12b534dd1bc90060791add4bf7ff0575
before: 0.077 after: 0.084 on cuda（device: 4080).
TORCH_COMPILE_DEBUG logs show that the fusion kernel of "triton_per_fused__softmax__to_copy_add_arange_embedding_exp_le_mul_native_dropout_permute_prepare_softmax_online_repeat_rsub_sub_unsqueeze_view_16" have been divided into 2 kernels: "triton_red_fused__to_copy_add_mul_permute_prepare_softmax_online_rsub_unsqueeze_view_12" + "triton_poi_fused__softmax__to_copy_add_exp_mul_native_dropout_permute_sub_unsqueeze_view_13"

reproducer:
python benchmarks/dynamo/huggingface.py --training --amp --amp-dtype float16 -d cuda -n10 --performance --disable-cudagraphs --only T5ForConditionalGeneration --cold-start-latency --backend=inductor

### Versions

torch                    2.11.0.dev20260105+cu128
triton                   3.6.0+git9844da95

cc @jerryzh168 @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #183881

Keep T5-style online softmax epilogues fused when their non-contiguous pointwise reads are compatible recomputations of reduction inputs, while preserving fully non-contiguous softmax behavior.

Fixes #177118

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the cleanup nits. I kept the non-INNER hint upgrade because narrowing it to DEFAULT breaks the T5 fusion regression test and prevents the intended persistent fusion.

File: torch/_inductor/codegen/simd_kernel_features.py [hunk 1]
```diff
@@ -197,20 +208,131 @@ def buffer_read_counts(self) -> dict[str, int]:
...
+    ) -> bool:
+        pointwise_nodes = list(pointwise_nodes)
+        if not pointwise_nodes:
+            return False
+
+        reduction_reads = [
+            dep
+            for node in reduction_nodes
+            if cls._is_online_softmax_reduction(node)
+            for dep in node.read_writes.reads
```
Comment (eellison) [inline torch/_inductor/codegen/simd_kernel_features.py]: LIkely, if we have a specific check for online softmax, the fusion is not being done in a general way

**Review by eellison — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/simd_kernel_features.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/codegen/simd_kernel_features.py]: My agent says I generalized the recomputed-read compatibility check so it is no longer tied to online softmax. The update keeps the conservative matching, coalesced-read, and non-contiguous-write guards while preserving the T5 fusion regression coverage.

File: torch/_inductor/codegen/simd_kernel_features.py [hunk 2]
```diff
@@ -197,20 +207,121 @@ def buffer_read_counts(self) -> dict[str, int]:
 
         return dict(read_counts)  # Convert defaultdict to regular dict
 
-    def has_non_contiguous_pw_in_reduction_kernel(self) -> bool:
-        pointwise_nodes = [
+    @staticmethod
+    def _deps_match(dep1: MemoryDep, dep2: MemoryDep) -> bool:
```
Comment (eellison) [inline torch/_inductor/codegen/simd_kernel_features.py]: this sort of deps match for a single change is usually suspect

**Review by eellison — COMMENTED** [review]
defering review on perf speedup until ive done the other reviews.

File: torch/_inductor/codegen/simd_kernel_features.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/codegen/simd_kernel_features.py]: My agent says I removed the custom normalized dependency matcher and now only treat an epilogue read as recomputed when the exact MemoryDep is already present in the reduction reads. The T5 fusion regression and the transposed-softmax negative test still pass with that narrower check.

Comment (eellison) [reply torch/_inductor/codegen/simd_kernel_features.py]: Can you explain more, on, what were the TORCH_LOGS="ir_pre_fusion, fusion, ir_post_fusion" and, why every change here was needed ?  what were the mismatching deps, what was the underyling problem, why do we have tiling issues ehre?

Comment (jansel) [reply torch/_inductor/codegen/simd_kernel_features.py]: My agent says the relevant logs were from the T5 attention softmax group: pre-fusion had an online_softmax_reduce over the S dimension and an epilogue that recomputed scores + position_bias before dropout/scaling. The position_bias access was the same MemoryDep already read by the reduction, but it is non-contiguous because T5 stores it with head as the innermost stride. Fusion then failed in two places: the kernel feature pass downgraded/kept the reduction away from the persistent inner path because it saw a non-contiguous pointwise read in the epilogue, and the scheduler pointwise+reduction tiling check also rejected the fused reduction because standalone pointwise tiling prefers that non-contiguous layout. The result in output_code was the reported split triton_red prepare_softmax_online kernel plus a separate triton_poi softmax/dropout epilogue.

The changes are needed as a pair: has_non_contiguous_pw_in_reduction_kernel now ignores only exact recomputed reads that are already reduction reads, which lets the persistent/INNER heuristic see this as a coalesced reduction instead of a genuinely bad pointwise epilogue; the can_fuse relaxation handles the later scheduler tiling gate for the same exact case. The guards are meant to keep this narrow: there must be a coalesced reduction read, there must be at least one exact recomputed read, unrelated non-contiguous reads still fail, non-contiguous writes still fail, and the transposed-softmax test covers the case we should not force into the persistent path.

**Comment by weishi-deng** [comment]
Hi @jansel, do we have a further fix or a fix plan for this PR and issue https://github.com/pytorch/pytorch/issues/177118? Thanks.

**Comment by jansel** [comment]
@weishi-deng just waiting for review on this one.  @eellison can you take another look?

**Comment by eellison** [comment]
this is fixed on main by https://github.com/pytorch/pytorch/pull/183521.
